lib/Twig/Extension/Core.php

```diff
@@ -821,7 +821,27 @@ function _twig_default_filter($value, $default = '')
 function twig_get_array_keys_filter($array)
 {
     if ($array instanceof Traversable) {
-        return array_keys(iterator_to_array($array));
+        while ($array instanceof IteratorAggregate) {
+            $array = $array->getIterator();
+        }
+
+        if ($array instanceof Iterator) {
+            $keys = array();
+            $array->rewind();
+            while ($array->valid()) {
+                $keys[] = $array->key();
+                $array->next();
+            }
+
+            return $keys;
+        }
```

Comment:
If you've hit this problem with real-world iterators, this seems reasonable. I just wouldn't have expected that iterators that have both useful keys and a key/value asymmetry occur sufficiently commonly in practice to merit this special-casing.

There is an RFC (https://wiki.php.net/rfc/foreach_void) for a `foreach ($it as $key => void)` syntax. IIRC this RFC never went to discussion, due to bad reception OTR, but maybe this is a good example to reconsider it.